scheds/include/scx/sdt_task_impl.bpf.h

```diff
@@ -0,0 +1,579 @@
+}
+
+static SDT_TASK_FN_ATTRS
+void sdt_task_free_to_pool(void __arena *ptr, struct sdt_task_pool __arena *pool)
+{
+	arena_list_node_t *elem;
+	__u64 __arena *data;
+	int i;
+
+	elem = arena_container_of(ptr, struct arena_list_node, data);
+
+	/* Zero out one word at a time since we cannot use memset. */
+	data = (__u64 __arena *)&elem->data;
+	cast_kern(data);
+
+	bpf_for(i, 0, pool->elem_size / 8) {
+		data[i] = (__u64)0;
+	}
+
+	bpf_spin_lock(&sdt_task_pool_alloc_lock);
+	list_add_head(elem, &pool->head);
+	bpf_spin_unlock(&sdt_task_pool_alloc_lock);
+}
+
+/* alloc desc and chunk and link chunk to desc and return desc */
+static SDT_TASK_FN_ATTRS struct sdt_task_desc __arena *sdt_alloc_chunk(void)
+{
+	struct sdt_task_chunk __arena *chunk;
+	struct sdt_task_desc __arena *desc;
+	struct sdt_task_desc __arena *out;
+
+	chunk = sdt_task_alloc_from_pool(&sdt_task_chunk_pool);
+	if (!chunk) {
+		bpf_printk("%s: failed to allocated chunk", __func__);
+		return NULL;
+	}
+
+	desc = sdt_task_alloc_from_pool(&sdt_task_desc_pool);
+	if (!desc) {
+		sdt_task_free_to_pool(chunk, &sdt_task_chunk_pool);
+		bpf_printk("%s: failed to allocated desc", __func__);
+		return NULL;
+	}
+
+	out = desc;
+
+	cast_kern(desc);
+
+	desc->nr_free = SDT_TASK_ENTS_PER_CHUNK;
+	desc->chunk = chunk;
+
+	return out;
+}
+
+static SDT_TASK_FN_ATTRS int sdt_pool_set_size(struct sdt_task_pool __arena *pool, __u64 data_size)
+{
+	/* All allocations are wrapped in a linked list node. */
+	data_size += sizeof(struct arena_list_node);
+
+	if (data_size > (SDT_TASK_ENT_PAGES * PAGE_SIZE)) {
+		bpf_printk("allocation size %ld too large", data_size);
+		return -E2BIG;
+	}
+
+	cast_kern(pool);
+	pool->elem_size = data_size;
+
+	return 0;
+}
+
+static SDT_TASK_FN_ATTRS int sdt_pool_set_size_data(struct sdt_task_pool __arena *pool, __u64 data_size, __u64 free_size)
+{
+	int ret;
+
+	ret = sdt_pool_set_size(pool, data_size);
+	if (ret)
+		return ret;
+
+	pool->free_size = free_size;
+
+	return 0;
+}
+
+/* initialize the whole thing, maybe misnomer */
+static SDT_TASK_FN_ATTRS int sdt_task_init(__u64 data_size)
+{
+	__u64 free_size;
+	int ret;
+
+	ret = sdt_pool_set_size(&sdt_task_chunk_pool, sizeof(struct sdt_task_chunk));
+	if (ret != 0)
+		return ret;
+
+	ret = sdt_pool_set_size(&sdt_task_desc_pool, sizeof(struct sdt_task_desc));
+	if (ret != 0)
+		return ret;
+
+	/* Page align and wrap data into a descriptor. */
+	data_size = div_round_up(data_size, 8) * 8;
+	free_size = data_size;
+	data_size += sizeof(struct sdt_task_data);
+
+	ret = sdt_pool_set_size_data(&sdt_task_data_pool, data_size, free_size);
+	if (ret != 0)
+		return ret;
+
+	sdt_task_desc_root = sdt_alloc_chunk();
+	if (sdt_task_desc_root == NULL)
+		return -ENOMEM;
+
+	return 0;
+}
+
+static SDT_TASK_FN_ATTRS
+int sdt_set_idx_state(struct sdt_task_desc __arena *desc, __u64 pos, bool state)
+{
+	__u64 __arena *allocated = desc->allocated;
+	__u64 bit;
+
+	cast_kern(allocated);
+
+	if (pos >= SDT_TASK_ENTS_PER_CHUNK) {
```

Comment:
`unlikely()`?

Reply:
I have added unlikely() on all the sanity checks that should never fail under normal operation.